# Assignment 5

Complete all numbered questions below. The math library is required, so it is imported here:

In [1]:
import math

## Section 1 -- Data Structures

#### Answer each of the following questions about data structures.

1. Do all the elements of a list have to be the same data type? What about tuples, dictionaries, and sets?

In [2]:
# They can all be mixed data types. No need to answer this with a code chunk.
L = [1,2.0,'a']
T = (1,2.0,'a')
D = {'k1':1,'k2':2.0,'k3':'a'}
S = {1,2.0,'a'}

print(L,T,D,S)

[1, 2.0, 'a'] (1, 2.0, 'a') {'k1': 1, 'k2': 2.0, 'k3': 'a'} {1, 2.0, 'a'}


2. Can a dictionary contain a list? Can a list contain a dictionary?

In [3]:
# Yes to both; prove using the lists and dictionaries defined in previous code chunk
new_dict = {'k1':1, 'k2':2.0, 'k3':L}
new_list = [1, 2.0, D]

print(new_dict, new_list)

{'k1': 1, 'k2': 2.0, 'k3': [1, 2.0, 'a']} [1, 2.0, {'k1': 1, 'k2': 2.0, 'k3': 'a'}]


3. Consider this small subset of data from the Star Wars dataset. Using the data structures you've learned, create an appropriate data structure to store this information.

| name           | height | mass  | eye_color |
|----------------|--------|-------|-----------|
| Luke Skywalker | 172    | 77.0  | blue      |
| C-3PO          | 167    | 75.0  | yellow    |
| R2-D2          | 96     | 32.0  | red       |
| Darth Vader    | 202    | 136.0 | yellow    |



In [4]:
# Mostly just designed to get them to think about combining data types to build their own dataframes.
# There are several ways to do this. Probably the best is a dictionary of lists:
starwars1 = {
            'name': ['Luke Skywalker', 'C-3PO', 'R2-D2', 'Darth Vader'],
            'height': [172, 167, 96, 202],
            'mass': [77.0, 75.0, 32.0, 136.0],
            'eye_color': ['blue', 'yellow', 'red', 'yellow']
            }

starwars1

# A dictionary of dictionaries could also make sense, but this doesn't fit with "tidy data"
starwars1 = {
            'Luke Skywalker': {'height': 172, 'mass': 77.0, 'eye_color': 'blue'},
            'C-3PO': {'height': 167, 'mass': 75.0, 'eye_color': 'yellow'} 
            }

# I omitted some lines here. A list of lists would also be fine.

## Section 2 - Operators

#### 1. For x, y, z, and w (defined below), write an expression to verify that the following statements are True (write your expression such that "True" is returned):

In [5]:
x = 1
y = [1,2,3]
z = [1,2,3]
w = [2,3,4]

#### (a) y and z are equal in value

In [6]:
y == z

True

#### (b) y and z are different objects

In [7]:
y is not z

True

#### (c) x is a member of y

In [8]:
x in y

True

#### (d) The value of x is not in the list w.

In [9]:
x not in w

True

#### (e) x is either a member of y or a member of w.

In [10]:
x in w or x in w

False

#### (f) x is a member of y and is not a member of w.

In [11]:
(x in y) and (x not in w)

True

## Section 3 -- Control Flow

#### 1. Create an empty list called "L". In a for loop, fill it with all numbers less than 100 that are divisible by 5 but not by 3. (Note : the list should not include 100)

In [12]:
L = []
for i in range(100):
    if i%5==0 and i%3!=0:
        L.append(i)
print(L)

[5, 10, 20, 25, 35, 40, 50, 55, 65, 70, 80, 85, 95]


#### 2. Do the same thing, this time using a while loop. (Note : make sure you re-initialize "L" to be empty!)

In [13]:
L = []
i = 0
while i<100:
    if i%5==0 and i%3!=0:
        L.append(i)
    i+=1

print(L)

[5, 10, 20, 25, 35, 40, 50, 55, 65, 70, 80, 85, 95]


#### 3. Do the same thing, this time using list comprehension. (Note : this time you shouldn't need to initialize L!)

In [14]:
L = [i for i in range(100) if i%5==0 and i%3!=0]
print(L)

[5, 10, 20, 25, 35, 40, 50, 55, 65, 70, 80, 85, 95]


#### 4. Make three lists L, M, and N. Using a for loop and an if-elif-else clause, fill L with all numbers 0-4 (inclusive), M with numbers 5-9 (inclusive), and N with any other numbers smaller than 15.

In [15]:
L = []
M = []
N = []

for i in range(15):
    if i<5:
        L.append(i)
    elif i<10:
        M.append(i)
    else:
        N.append(i)
        
print(L, M, N)

[0, 1, 2, 3, 4] [5, 6, 7, 8, 9] [10, 11, 12, 13, 14]


## Section 4 -- List Comprehension

#### 1. Using list comprehension, for all numbers less than 20, build a list that squares the even numbers and cubes the odd numbers.

In [16]:
L = [i**2 if i%2==0 else i**3 for i in range(20)]
print(L)

[0, 1, 4, 27, 16, 125, 36, 343, 64, 729, 100, 1331, 144, 2197, 196, 3375, 256, 4913, 324, 6859]


#### 2. Rebuild the same list with list comprehension, but only perform an operation if the original number is divisible by 3 (should get [0, 27, 36, 729, 144, 3375, 324].

In [17]:
L = [i**2 if i%2==0 else i**3 for i in range(20) if not i%3]
print(L)

[0, 27, 36, 729, 144, 3375, 324]


## Section 5 -- Functions

#### Test the following functions with L = list(range(10)) and x = 3, defined below. 

In [18]:
L = list(range(10))
x = 3

#### 1. Define a function that takes in a list L and a number x. Have it return a new list of only the elements of L that are divisible by x. Start simple with this function. We'll add to it below.

In [19]:
def divisible_by(L, x):
    return [i for i in L if not i%x]

L = list(range(10))
x = 3

N = divisible_by(L,x)
print(N)

[0, 3, 6, 9]


#### 2. Modify this function below so that it does not return 0 in the list.

In [20]:
def nonzero_divisible_by(L, x):
    return [i for i in L if not i%x and i]

L = list(range(10))
x = 3

N = nonzero_divisible_by(L,x)
print(N)

[3, 6, 9]


#### 3. This function will not work if L is not a list and x is not an integer. Redefine the function below and add a statement that checks to make sure the data types are correct.

In [21]:
def nonzero_divisible_by(L, x):
    if type(L) is not list:
        raise TypeError("L must be list")
    if type(x) is not int:
        raise TypeError("x must be integer")
        
    return [i for i in L if not i%x and i]

L = list(range(10))
x = 3

nonzero_divisible_by(L,x)
# nonzero_divisible_by(L,0.3)
# nonzero_divisible_by("h",x)

[3, 6, 9]

#### 4. Write both_na(), a function that takes two vectors of the same length and returns the number of positions that have an NA in both vectors. You'll need the math library for this part (Note: NAN is math.nan). Test the function with x1, y1 and x2, y2, defined below.

In [22]:
x1 = [math.nan, math.nan, 1, 2]
y1 = [math.nan, 1, math.nan, 2]
x2 = [math.nan, math.nan, 1, 2, math.nan, math.nan, 1]
y2 = [math.nan, 1, math.nan, 2, math.nan, math.nan, 1]

In [23]:
def both_na(x, y):
    if len(x)!=len(y):
        raise ValueError("List lengths are not the same")
    L = len(x)
    count = 0
    for i,j in zip(x,y):
        if math.isnan(i) and math.isnan(j):
            count+=1
    return(count)
    
x1 = [math.nan, math.nan, 1, 2]
y1 = [math.nan, 1, math.nan, 2]
x2 = [math.nan, math.nan, 1, 2, math.nan, math.nan, 1]
y2 = [math.nan, 1, math.nan, 2, math.nan, math.nan, 1]

print(both_na(x1,y1), both_na(x2,y2))

1 3


#### 5. Create each of the functions as described in the space provided. Test each of the functions with the list x3, defined below.

In [24]:
x3 = [2, 5, 3, 10, 4, math.nan, math.nan]

(a) Create a function that returns the last value in a list. Test with x3.

In [25]:
def last_val(x):
    return x[-1]

(b) Create a function that returns the elements at even numbered positions. Test with x3.

In [26]:
def even_pos(x):
    return [val for i,val in enumerate(x) if not i%2]

(c) Create a function that returns every element except the last value. Test with x3.

In [27]:
def all_but_last(x):
    return x[:-1]

(d) Create a function that returns only even numbers (and no missing values).

In [28]:
def even_vals(x):
    return [i for i in x if ((not i%2) and (not math.isnan(i)))]

In [29]:
# Putting test in one code chunk down here

print(last_val(x3))
print(even_pos(x3))
print(all_but_last(x3))
print(even_vals(x3))

nan
[2, 3, 4, nan]
[2, 5, 3, 10, 4, nan]
[2, 10, 4]


#### 6. Define a function called "negate_evens" that takes in a number "x" and returns -x if x is even or x if x is odd. Use map to apply this function to numbers 1 - 10 (inclusive) and store the results in a list.

In [30]:
def negate_evens(x):
    if x%2:
        return x
    else:
        return -x

L = list(map(negate_evens, range(1,11)))
print(L)

[1, -2, 3, -4, 5, -6, 7, -8, 9, -10]
